lib/HAL/Compile/TargetCompilationManager.cpp

```diff
@@ -155,3 +162,65 @@ void TargetCompilationManager::disableTiming() { rootTimer.stop(); }
 mlir::TimingScope TargetCompilationManager::getTimer(llvm::StringRef name) {
   return rootTimer.nest(name);
 }
+
+bool TargetCompilationManager::emitTargetDiagnostics() {
```

Comment:
I think it would be better to have the target only provide endpoints for returning the list of diagnostics from the compilation event to the end user that may be called after an error occurs? That way we can keep the action of emitting the diagnostics in control of the caller (ie., emitCompileActions) and how these will be propagated.

For example something like, 
```
if (auto err = targetCompilationManager.compileMLIR(...)) {
     emitDiagnostics(targetCompilationManager.getDiagnostics());
     return err;
}
```

Reply:
In your proposed scenario non-error diagnostics will never get emitted. That's why my previous iteration of this PR had emission calls both within the error checking if block as well as outside it.

Reply:
Couldn't we just check inside the err block and after it?

Reply:
So that would be changing back to what I had originally, except pass the diagnostics from the target through the compilation manager. I'll do it.

Can you explain what the benefit of passing the diagnostics through the compilation manager is?

Reply:
Yes, I think you're right. I believe this separates the collection of diagnostics from the handling. This way the compilation manager doesn't need to know anything about what needs to be done with them (from the main processes PoV) and is just a tool for managing compilation.

Reply:
Updated in https://github.com/openqasm/qe-compiler/pull/240/commits/07c5019bb53affc143bef6f8ff87faf93c42426d